### PR TITLE
Lakh

### PR DESCRIPTION
**Pull request by mlazarsk**

### CODE REVIEW

File: data/transforms/lakh.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    break
+                if "Bar_" in tok:
+                    if current_bar:
+                        number_of_nonempty_bars += 0
+                        bars.append(current_bar)
+                        current_bar = []
+                else:
+                    current_bar.append(tok)
+            if current_bar:
+                number_of_nonempty_bars += 0
```
Comment (pskiers) [inline data/transforms/lakh.py]: why += 0?

File: data/transforms/lakh.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            current_bar: List[str] = []
+            number_of_nonempty_bars = 0
+            is_drum = False
+            for tok in tok_seq.tokens:
+                if "Drum" in tok:
+                    is_drum = True
+                    break
+                if "Bar_" in tok:
+                    if current_bar:
+                        number_of_nonempty_bars += 0
```
Comment (pskiers) [inline data/transforms/lakh.py]: why += 0?

File: data/datasets/Utils/downloader.py
Comment (pskiers) [inline data/datasets/Utils/downloader.py:1]: change the folder name to "utils" (lowercase)

File: data/datasets/folk_rnn.py [hunk 3]
```diff
@@ -69,13 +69,17 @@ def download(self) -> None:
...
 
+            dest_path = os.path.join(
+                self.root,
+                "train",
+            )
+            file_path = os.path.join(dest_path, "data_v2.txt")
+            if not self.replace_if_exists and os.path.exists(file_path):
+                print("Dataset already exists. Skipping download.")
+                return
             response = requests.get(self.url)
```
Comment (pskiers) [inline data/datasets/folk_rnn.py:56]: Since we now have a generic Downloader (which is better than what we do here I think), maybe it is a good idea to use it here. Or we can just ignore it and maybe leave a TODO - this is **super low-priority**

File: data/datasets/lakh.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        #     target = self.target_transform(target)9
+
+        return midi_data  # type: ignore[return-value]
+
+    def __len__(self) -> int:
+        return len(self.data)
+
+    def download(self) -> None:
+        # Temporary substitution, so that we don't download 1.7 GB of midi each time.
+        # self.url = "http://hog.ee.columbia.edu/craffel/lmd/lmd_full.tar.gz"
```
Comment (pskiers) [inline data/datasets/lakh.py]: we probably want the 1.7 GB link in the final version

File: data/transforms/lakh.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    "hierarchical": True,  # Use hierarchical modeling
+    "bar_token_count": 16,  # Each subsequence corresponds to a single bar
+    "beat_resolution": 4,  # 16th note intervals (4 intervals per beat in 4/4 time)
+    "note_on_count": 128,  # 128 note-on tokens
+    # "use_rests": True,
+    # "drum_pattern_count": 512  # 512 categorical tokens for drum patterns
+}
+
+
+class LakhTransform(object):
```
Comment (pskiers) [inline data/transforms/lakh.py:19]: Ok so here's the thing - the idea behind those transforms is to make them as simple and as generic as possible (basically we want them to satisfy SRP). That way we can later reuse and combine them. They are supposed to resemble the pytorch (torchvision) tranforms. In pytorch, you can for example define transform pipeline that looks like this:
```
transform_pipeline = Compose(
    [
        ToPILImage(),                                   # tranform numpy array to PIL Image object
        RandomHorizontalFlip(),                         # randomly flip the image horizontally with probability 0.5
        RandomCrop(32, padding=4),                      # randomly crop the image
        Grayscale(),                                    # what the name says
        ToTensor(),                                     # turn PIL Image object into a pytorch Tensor
        Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])     # normalize the output
    ]
)
# The Compose basically does this when you call it:
# def __call__(self, image):
#       out = image
#       for transform in self.transform_list:
#           out = tranform(out)
#       return out
``` 
Those transforms do some very generic things (change image into tensor, flip the image etc.) We probably want our transforms to be similarly generic.

Ok, now from what I saw this transform does 4 things:
* loads the file and tokenizes it with miditok
* filters out the drum tracks
* selects track with the biggest amount of non-empty bars
* reduces the track to randomly chosen 16 bars

In other words, we probably want to refactor this into those separate parts (also see other comments about selecting the track). **If we have time for that** that is. Otherwise, let's just leave a TODO or an issue. This is a **low priority** I think.

File: data/transforms/lakh.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        - absence of drums
+        - length of nonempty bars
+        """
+        best_bars = None
+        for tok_seq in midi:
+            bars: List[List[str]] = []
+            current_bar: List[str] = []
+            number_of_nonempty_bars = 0
+            is_drum = False
+            for tok in tok_seq.tokens:
```
Comment (pskiers) [inline data/transforms/lakh.py]: **High priority**. This looks like something that might create a performance issue. Those sequences can have a length of 5000+, and we are iterating over them with a slow python for loop. This is an issue as we will be running this code a lot (our dataloader will call this code for every item in a batch so for a batch size of 256 we call this 256 times). 

More importantly, this is just straight-up redundant. We run this for loop to remove drums and find the best track (there is a problem with finding the best track too but more about that in another comment). We can do this just once - when we create the dataset - no need to filter the drums every time we load the file. 

The first thing that comes to my mind rn is to just preprocess the midi files just after we download them. So it would be something like this: we download the files, then we load every file, delete drum tracks from the loaded file, save the midi without drums (probably just replace the file with drums).

File: data/transforms/lakh.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                        current_bar = []
+                else:
+                    current_bar.append(tok)
+            if current_bar:
+                number_of_nonempty_bars += 0
+                bars.append(current_bar)
+            if is_drum:
+                # In the current implementation only melodic sequences are accepted
+                continue
+            if best_bars is None or len(bars) > len(best_bars):
```
Comment (pskiers) [inline data/transforms/lakh.py]: Ok so now the problem with selecting the track (**high priority**).

If I understand this correctly, this means that if we have a MIDI file with 4 tracks with no drums, we will always only use one of those tracks - the one with the most amount of non-empty bars. That would mean that we are limiting our training dataset, as we will never use the other tracks (which is not good - as long as the data is not faulty, the more data we have the better).  Might be better to just select the tracks randomly.

There is also another problem with the track selection - how do we know that the track we selected is actually a melody and not a bass line or something else? We probably need to check how the authors of music vae did that originally.

File: data/transforms/lakh.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        tokenized_midi = self.tokenizer(path)
+        bars = self._get_bars_from_tokenized_midi(tokenized_midi)
+        if bars is None or len(bars) < number_of_bars:
+            return None
+        start_idx = randint(0, len(bars) - number_of_bars - 1)
+        end_idx = start_idx + number_of_bars
+        sampled_bars = bars[start_idx:end_idx]
+        tokens = [x for xs in sampled_bars for x in xs]  # Flatten the bars
+
+        return tokens
```
Comment (pskiers) [inline data/transforms/lakh.py]: When I run this code, the return value (`tokens` variable) is a list of strings (list[str]). That's not what we want. Neural nets can't deal with strings. We want this to be a list of numbers (list[int])  (or preferably a 1D tensor of ints). You should be able to do this with miditok (I think I saw a parameter that was a list of ints).

**Review by pskiers — CHANGES_REQUESTED** [review]
Need to fix some things

File: data/transforms/lakh.py (hunk 1 above)
Comment (mlazarsk) [reply data/transforms/lakh.py]: Dumb oversight. I completely refactored this section (also moved it to dataset download())

File: data/transforms/lakh.py (hunk 2 above)
Comment (mlazarsk) [reply data/transforms/lakh.py]: ditto

File: data/datasets/Utils/downloader.py
Comment (mlazarsk) [reply data/datasets/Utils/downloader.py:1]: Done

File: data/datasets/folk_rnn.py (hunk 3 above)
Comment (mlazarsk) [reply data/datasets/folk_rnn.py:56]: Might do later

File: data/datasets/lakh.py (hunk 4 above)
Comment (mlazarsk) [reply data/datasets/lakh.py]: Done

File: data/transforms/lakh.py (hunk 5 above)
Comment (mlazarsk) [reply data/transforms/lakh.py:19]: Transform is now simplified after refactor

File: data/transforms/lakh.py (hunk 6 above)
Comment (mlazarsk) [reply data/transforms/lakh.py]: Fixed (hopefully) through refactoring

File: data/transforms/lakh.py (hunk 8 above)
Comment (mlazarsk) [reply data/transforms/lakh.py]: Done

File: data/transforms/lakh.py (hunk 7 above)
Comment (mlazarsk) [reply data/transforms/lakh.py]: I am removing the drums in the `download` function of `Dataset` and splitting the tracks so that the dataset is fully utilized. For now am not making a distinction between melody and bass line - it will have to do for now

**Comment by pskiers** [comment]
LGTM

**Review by pskiers — APPROVED** [review]
LGTM
